contracts/azorius/LinearERC721Voting.sol

```diff
@@ -0,0 +1,292 @@
+    error InvalidVote();
+    error InvalidTokenAddress();
+    error NoVotingWeight();
+    error TokenAlreadySet();
+    error LastGovernanceToken();
+
+    function setUp(bytes memory initializeParams) public override initializer {
+        (
+            address _owner,
+            GovernanceNFT[] memory _governanceTokens, // TODO split this up into parameters and use addGovernanceToken instead?
+            address _azoriusModule,
+            uint32 _votingPeriod,
+            uint256 _quorumThreshold,
+            uint256 _basisNumerator
+        ) = abi.decode(
+            initializeParams,
+            (address, GovernanceNFT[], address, uint32, uint256, uint256)
+        );
+
+        // TODO see above
+        for (uint i = 0; i < _governanceTokens.length; i++) {
+            governanceTokens.push(_governanceTokens[i]);
+        }
+
+        __Ownable_init();
+        transferOwnership(_owner);
+        _setAzorius(_azoriusModule);
+        _updateQuorumThreshold(_quorumThreshold);
+        _updateBasisNumerator(_basisNumerator);
+        _updateVotingPeriod(_votingPeriod);
+
+        emit StrategySetUp(_azoriusModule, _owner);
+    }
+
+    function updateVotingPeriod(uint32 _votingPeriod) external onlyOwner {
+        _updateVotingPeriod(_votingPeriod);
+    }
+
+    function updateQuorumThreshold(uint256 _quorumThreshold) external onlyOwner {
+        _updateQuorumThreshold(_quorumThreshold);
+    }
+
+    function getProposalVotes(uint32 _proposalId) external view
+        returns (
+            uint256 noVotes,
+            uint256 yesVotes,
+            uint256 abstainVotes,
+            uint32 startBlock,
+            uint32 endBlock
+        )
+    {
+        noVotes = proposalVotes[_proposalId].noVotes;
+        yesVotes = proposalVotes[_proposalId].yesVotes;
+        abstainVotes = proposalVotes[_proposalId].abstainVotes;
+        startBlock = proposalVotes[_proposalId].votingStartBlock;
+        endBlock = proposalVotes[_proposalId].votingEndBlock;
+        // TODO why didn't we return has voted data in ERC20 strategy?
+    }
+
+    // voting requires providing the NFT addresses and ids, as IERC721 does not have a method
+    // for determining which NFT ids a particular address holds 
+    function vote(uint32 _proposalId, uint8 _support, bytes memory _nftData) external {
+        ( 
+            address[] memory tokenAddresses,
+            uint256[] memory tokenIds  
+        ) = abi.decode(_nftData, (address[], uint256[]));
+
+        uint256 weight = getWeight(_proposalId, tokenAddresses, tokenIds, msg.sender);
+
+        if (weight == 0) revert NoVotingWeight();
+
+        _vote(_proposalId, msg.sender, _support, weight);
+    }
+
+    // verifies the voter holds the NFTs and returns the total weight associated with their tokens
+    // the frontend will need to determine whether an address can vote on a proposal, as it is possible
+    // to vote twice if you get more weight later on
+    function getWeight(
```

Comment:
I would think we'd want this function to be available as a `public view` function, and not changing state in any way. To avoid looping through all the tokenAddresses / IDs in two places, maybe this function only takes a single `tokenAddress` and `tokenId`, (and `proposalId & `voter) and then just returns the weight